added retry logic to HTTP requests

jameslamb opened the pull request:
Thanks for this awesome project!

In this PR, I'd like to propose swapping out calls to `httr::GET()`, `httr::PATCH()`, etc. with `httr::RETRY()`. This will make the package more resilient to transient problems like brief network outages or periods where the service(s) it hits are overwhelmed. In my experience, using retry logic almost always improves the user experience with HTTP clients.

I apologize for the whitespace changes. I have my editor set to trim excess horizontal whitespace when I save files. If you'd like me to revert those please let me know.

cpsievert commented:
Thank you for the PR, and sorry for delayed response!

That'd be great if you could revert the whitespace changes. Also, I think you missed [this call to `httr::VERB()`](https://github.com/ropensci/plotly/blob/9bcf90a580faa96f53faa5009a9cab435dc40a25/R/api_exports.R#L190)

jameslamb commented:
Thanks. I've reverted all of the whitespace changes and replaced the `httr::VERB()`

jameslamb commented on R/api_exports.R:
```diff
@@ -187,9 +188,16 @@ api <- function(endpoint = "/", verb = "GET", body = NULL, ...) {
+    , api_auth()
+    , body = body
+    , times = 5
+    , terminate_on = c(400, 401, 403, 404)
```
In my experience, these particular statuses tend not to be resolved by retries. If you have a preferred set (including none), I'd be happy to change it. My main goal with this PR is to switch from one-shot all-or-nothing requests to retries. I don't have a strong opinion about `terminate_on`.

cpsievert commented on R/api_exports.R:
```diff
@@ -170,6 +170,7 @@ api_download_file <- function(id, username, endpoint = "files", ...) {
 
 
 #' @rdname api
+#' @importFrom httr RETRY
```
Please remove this line as it's also in `R/imports.R`

cpsievert commented on R/plotly_IMAGE.R:
```diff
@@ -9,7 +9,8 @@
 #' @param out_file A filename for writing the image to a file.
-#' @param ... arguments passed onto `httr::POST`
+#' @param ...  arguments passed onto `httr::RETRY`
+#' @importFrom httr RETRY write_disk
```
Please remove this line as it's also in `R/imports.R`

jameslamb replied:
oh I see, my mistake. I've not seen that pattern of putting all `importFrom`s in one place before. I'll remove this

cpsievert commented:
Sorry for the delay in getting this merged @jameslamb, this is looking good, and I'd be happy to take it on. Just a couple minor nitpicks:

* `terminate_on` was added in [httr 1.3.0](https://github.com/r-lib/httr/blob/master/NEWS.md#httr-130), so lets do `httr (>= 1.3.0)` in the `DESCRIPTION` file.
* Please use trailing commas (instead of starting a new line with a comma).

jameslamb commented:
> Sorry for the delay in getting this merged @jameslamb, this is looking good, and I'd be happy to take it on. Just a couple minor nitpicks:
> 
> * `terminate_on` was added in [httr 1.3.0](https://github.com/r-lib/httr/blob/master/NEWS.md#httr-130), so lets do `httr (>= 1.3.0)` in the `DESCRIPTION` file.
> * Please use trailing commas (instead of starting a new line with a comma).

will do, thanks for the review!

jameslamb commented:
Ok I  just made the requested changes in https://github.com/ropensci/plotly/pull/1656/commits/1bf6a4a09bfa0bb39270d356862af0555a92f6f9. I've also rebased  to `master` to catch changes that have happened since this was opened.

cpsievert commented:
Awesome, thank you! One last request: could you add a NEWS.md item (under improvements)?

jameslamb commented:
Sure! Just added in https://github.com/ropensci/plotly/pull/1656/commits/0f1f1819dc18eac493cc85d784e49afc524f377a

I also fixed a missing comma from my last commit, which I think was the reason Travis was ❌
